heartbeat/monitors/active/http/respbody.go

```diff
@@ -0,0 +1,126 @@
+	"fmt"
+	"io"
+	"net/http"
+	"unicode/utf8"
+
+	"github.com/elastic/beats/heartbeat/eventext"
+	"github.com/elastic/beats/libbeat/common"
+
+	"github.com/elastic/beats/heartbeat/reason"
+	"github.com/elastic/beats/libbeat/beat"
+)
+
+func handleRespBody(event *beat.Event, resp *http.Response, responseConfig responseConfig, errReason reason.Reason) error {
+	defer resp.Body.Close()
+
+	sampleMaxBytes := responseConfig.IncludeBodyMaxBytes
+
+	includeSample := responseConfig.IncludeBody == "always" || (responseConfig.IncludeBody == "on_error" && errReason != nil)
+
+	// No need to return any actual body bytes if we'll discard them anyway. This should save on allocation
+	if !includeSample {
+		sampleMaxBytes = 0
+	}
+
+	sampleStr, bodyBytes, bodyHash, err := readResp(resp, sampleMaxBytes)
+	if err != nil {
+		return err
+	}
+
+	evtBodyMap := common.MapStr{
+		"hash":  bodyHash,
+		"bytes": bodyBytes,
+	}
+	if includeSample {
+		evtBodyMap["content"] = sampleStr
+	}
+
+	eventext.MergeEventFields(event, common.MapStr{
+		"http": common.MapStr{
+			"response": common.MapStr{"body": evtBodyMap},
+		},
+	})
+
+	return nil
+}
+
+// readResp reads the first sampleSize bytes from the httpResponse,
+// then closes the body (which closes the connection). It doesn't return any errors
+// but does log them. During an error case the return values will be (nil, -1).
+// The maxBytes params controls how many bytes will be returned in a string, not how many will be read.
+// We always read the full response here since we want to time downloading the full thing.
+// This may return a nil body if the response is not valid UTF-8
+func readResp(resp *http.Response, maxSampleBytes int) (bodySample string, bodySize int64, hashStr string, err error) {
+	if resp == nil {
+		return "", -1, "", fmt.Errorf("cannot readResp of nil HTTP response")
+	}
+
+	respSize, bodySample, hash, err := readPrefixAndHash(resp.Body, maxSampleBytes)
+
+	return bodySample, respSize, hash, err
+}
+
+func readPrefixAndHash(body io.ReadCloser, maxPrefixSize int) (respSize int64, prefix string, hashStr string, err error) {
+	hash := sha256.New()
+	// Function to lazily get the body of the response
+	rawBuf := make([]byte, 1024)
+
+	// Buffer to hold the prefix output along with tracking info
+	prefixBuf := make([]byte, maxPrefixSize)
+	prefixRemainingBytes := maxPrefixSize
+	prefixWriteOffset := 0
+	for {
+		readSize, readErr := body.Read(rawBuf)
+
+		respSize += int64(readSize)
+		hash.Write(rawBuf[:readSize])
+
+		if prefixRemainingBytes > 0 {
+			if readSize >= prefixRemainingBytes {
+				copy(prefixBuf[prefixWriteOffset:maxPrefixSize], rawBuf[:prefixRemainingBytes])
+				prefixWriteOffset += prefixRemainingBytes
+				prefixRemainingBytes = 0
+			} else {
+				copy(prefixBuf[prefixWriteOffset:prefixWriteOffset+readSize], rawBuf[:readSize])
+				prefixWriteOffset += readSize
+				prefixRemainingBytes -= readSize
+			}
+		}
+
+		if readErr == io.EOF {
+			break
+		}
+
+		if readErr != nil {
```

Comment:
Sorry, just seeing this now. Shouldn't this line directly be after line 95? I get that if you have an EOF that you also want to get the content first, but if you have an actual error, I assume also the readSize is probably 0 ?

Reply:
Thanks for the reminder 👍